presto-common/src/main/java/com/facebook/presto/common/zorder/ZOrderIntegers.java

```diff
@@ -0,0 +1,98 @@
+ */
+package com.facebook.presto.common.zorder;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class ZOrderIntegers
+{
+    private ZOrderIntegers() {}
+
+    // TODO: make int and long z-address version
+    public static byte[] encode(List<Integer> input, List<Integer> bitRanges)
+            throws Exception
+    {
+        if (input.size() > 10) {
+            throw new Exception("Z-ordering does not support more than 10 input numbers");
+        }
+
+        // Transfer into array for faster access and to change values
+        int[] inputList = new int[input.size()];
+        for (int index = 0; index < input.size(); index++) {
+            inputList[index] = input.get(index);
+            if (inputList[index] < 0) {
+                throw new Exception("Negative numbers are not supported for z-ordering");
+            }
+        }
+
+        int totalBitLength = 0;
+        int maxBitLength = 0;
+        for (int bitLength : bitRanges) {
+            totalBitLength += bitLength;
+            maxBitLength = Math.max(maxBitLength, bitLength);
+        }
+
+        byte[] address = new byte[(totalBitLength + 7) >> 3];
+        int bitIndex = totalBitLength - 1;
+        int bitOffset = bitIndex & 7;
+        // Interweave input bits into address from the most significant bit to preserve data locality
+        for (int inputBitPos = 0; inputBitPos < maxBitLength; inputBitPos++) {
+            for (int index = 0; index < input.size(); index++) {
+                if (inputBitPos >= bitRanges.get(index)) {
+                    continue;
+                }
+                int maskedBit = (inputList[index] >> (bitRanges.get(index) - inputBitPos - 1)) & 1;
+                address[bitIndex >> 3] |= maskedBit << bitOffset;
+                bitIndex--;
+                bitOffset = bitIndex & 7;
+            }
+        }
+        return address;
+    }
+
+    public static List<Integer> decode(List<Integer> bitRanges, byte[] address)
+    {
+        int listLength = bitRanges.size();
+        int[] output = new int[listLength];
+
+        int totalBitLength = 0;
+        int maxBitLength = 0;
+        for (int bitLength : bitRanges) {
+            totalBitLength += bitLength;
+            maxBitLength = Math.max(maxBitLength, bitLength);
+        }
+
+        int bitIndex = totalBitLength - 1;
+        int bitOffset = bitIndex & 7;
+        // Un-weave address into original integers
+        for (int outputBitPos = 0; outputBitPos < maxBitLength; outputBitPos++) {
+            for (int index = 0; index < listLength; index++) {
+                if (outputBitPos >= bitRanges.get(index)) {
+                    continue;
+                }
+                int maskedBit = (address[bitIndex >> 3] >> bitOffset) & 1;
+                output[index] |= maskedBit << (bitRanges.get(index) - outputBitPos - 1);
+                bitIndex--;
+                bitOffset = bitIndex & 7;
+            }
+        }
```

Comment:
Interesting. Can you see if the following logic is simpler?

int bitPos = 0;
while (bitIndex > 0) {
    for (int i = 0; i < listLength; i++) {
        if (bitPos >= bitRanges.get(i)) continue;
       set the bit to the number.
       bitIndex--;
    }
   bitPos++;
}